pkg/providers/v2/instances_test.go

```diff
@@ -0,0 +1,216 @@
+
+	return &ec2.DescribeInstancesOutput{
+		Reservations: []*ec2.Reservation{
+			{
+				Instances: []*ec2.Instance{
+					&instance,
+				},
+			},
+		},
+	}, nil
+}
+
+func newFakeInstances() (cloudprovider.InstancesV2, error) {
+	return &instances{
+		ec2: fakeEC2{},
+	}, nil
+}
+
+func makeInstance(num int, privateIP, publicIP, privateDNSName, publicDNSName string, setNetInterface bool, stateName string) ec2.Instance {
+	instance := ec2.Instance{
+		InstanceId:       aws.String(fmt.Sprintf("i-%d", num)),
+		PrivateDnsName:   aws.String(privateDNSName),
+		PrivateIpAddress: aws.String(privateIP),
+		PublicDnsName:    aws.String(publicDNSName),
+		PublicIpAddress:  aws.String(publicIP),
+		InstanceType:     aws.String("c3.large"),
+		Placement:        &ec2.Placement{AvailabilityZone: aws.String("us-west-1a")},
+		State: &ec2.InstanceState{
+			Name: aws.String("running"),
+		},
+	}
+	if setNetInterface == true {
+		instance.NetworkInterfaces = []*ec2.InstanceNetworkInterface{
+			{
+				Status: aws.String(ec2.NetworkInterfaceStatusInUse),
+				PrivateIpAddresses: []*ec2.InstancePrivateIpAddress{
+					{
+						PrivateIpAddress: aws.String(privateIP),
+					},
+				},
+			},
+		}
+	}
+	return instance
+}
+
+func makeNode(nodeName string, offset int) *v1.Node {
+	id := uuid.New()
+	instanceID := fmt.Sprintf("i-%v-%d", id.String(), offset)
+	instance := &ec2.Instance{}
+	instance.InstanceId = aws.String(instanceID)
+	instance.Placement = &ec2.Placement{
+		AvailabilityZone: aws.String("us-west-1a"),
+	}
+	instance.PrivateDnsName = aws.String(fmt.Sprintf("ip-172-20-0-%d.ec2.internal", 101+offset))
+	instance.PrivateIpAddress = aws.String(fmt.Sprintf("192.168.0.%d", 1+offset))
+
+	var tag ec2.Tag
+	tag.Key = aws.String("KubernetesCluster")
+	tag.Value = aws.String(TestClusterID)
+	instance.Tags = []*ec2.Tag{&tag}
+
+	providerID := "aws://us-west-1a/" + instanceID
+	return &v1.Node{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: nodeName,
+		},
+		Spec: v1.NodeSpec{
+			ProviderID: providerID,
+		},
+	}
+}
+
+func TestGetInstanceProviderID(t *testing.T) {
+	instance0 := makeInstance(0, "192.168.0.1", "1.2.3.4", "instance-same.ec2.internal", "instance-same.ec2.external", true, "running")
+	instance1 := makeInstance(1, "192.168.0.2", "", "instance-same.ec2.internal", "", false, "running")
+	instance2 := makeInstance(2, "192.168.0.1", "1.2.3.4", "instance-other.ec2.internal", "", false, "running")
+
+	testCases := []struct {
+		instance   ec2.Instance
+		providerID string
+	}{
+		{instance0, "aws://us-west-1a/i-0"},
+		{instance1, "aws://us-west-1a/i-1"},
+		{instance2, "aws://us-west-1a/i-2"},
+	}
+
+	for _, testCase := range testCases {
+		providerID, err := getInstanceProviderID(&testCase.instance)
+		assert.NoError(t, err)
+		assert.Equal(t, testCase.providerID, providerID)
+	}
+}
+
+func TestInstanceExists(t *testing.T) {
```

Comment:
@andrewsykim Updated the unit tests, let me know if you have any other suggestions :)